### PR TITLE
Compatibility with PHP 8.1

### PR DESCRIPTION
**Pull request by srjlewis**
Add typing and compatibility with PHP 8.1

I have also pulled in #191

### CODE REVIEW

**Comment by kenjis** [comment]
@koriym Can you review?

File: src/ConnectionLocator.php [hunk 1]
```diff
@@ -42,21 +42,21 @@ class ConnectionLocator implements ConnectionLocatorInterface
...
      */
-    protected $write = [];
+    protected array $write = [];
 
     /**
      *
      * Constructor.
      *
-     * @param callable $default A callable to create a default connection.
+     * @param callable|null $default A callable to create a default connection.
```
Comment (koriym) [inline src/ConnectionLocator.php:51]: Isn't "?callable" good enough?

File: composer.json [hunk 2]
```diff
@@ -22,8 +22,9 @@
         }
     ],
     "require": {
-        "php": ">=5.6.0",
-        "psr/log": "^1.0"
+        "php": ">=8.0",
```
Comment (kenjis) [inline composer.json]: Aura 5.x requires PHP 8.1 or later.

```suggestion
        "php": ">=8.1",
```

Comment (srjlewis) [reply composer.json]: It does work fine on php 8.0, but I am fine with it being set to 8.1, ~~did you mean to remove the psr/log requirement?~~ sorry miss read the change

Comment (kenjis) [reply composer.json]: Aura 5.x does not support PHP 8.0.
PHP 8.0 is supported by Aura 4.x or 3.x.

I did not mean remove the psr/log.

File: .github/workflows/tests.yml [hunk 3]
```diff
@@ -5,7 +5,7 @@ jobs:
     runs-on: ubuntu-latest
     strategy:
       matrix:
-        php-versions: ['5.6', '7.0', '7.1', '7.2', '7.3', '7.4', '8.0']
+        php-versions: ['8.0', '8.1']
```
Comment (kenjis) [inline .github/workflows/tests.yml]:
```suggestion
        php-versions: ['8.1']
```

Comment (srjlewis) [reply .github/workflows/tests.yml]: I was just about do that change.

**Review by koriym — APPROVED** [review]

File: src/ConnectionLocator.php (hunk 1 above)
Comment (srjlewis) [reply src/ConnectionLocator.php:51]: nullable notation `?<type>` is not recognized within docblock by static analyzers such as phpStan.

phpStan Docs https://phpstan.org/writing-php-code/phpdoc-types
phpdoc guide https://docs.phpdoc.org/3.0/guide/guides/types.html

So I am using union type notation `<type>|null` as shown in the above docs.

**Review by srjlewis — COMMENTED** [review]
Add a extra test, all looks good

File: src/ConnectionLocator.php (hunk 1 above)
Comment (koriym) [reply src/ConnectionLocator.php:51]: Got it. Thanks for the clarification.

File: src/PdoInterface.php [hunk 4]
```diff
@@ -185,5 +186,5 @@ public function setAttribute($attribute, $value);
      * @return array
      *
      */
-    public static function getAvailableDrivers();
+    //public static function getAvailableDrivers(): array;
```
Comment (kenjis) [inline src/PdoInterface.php]: Why commented out?

File: src/PdoInterface.php [hunk 5]
```diff
@@ -70,18 +70,18 @@ public function errorInfo();
      * @see http://php.net/manual/en/pdo.exec.php
      *
      */
-    public function exec($statement);
+    public function exec(string $statement): int;
```
Comment (kenjis) [inline src/PdoInterface.php]:
```suggestion
    public function exec(string $statement): int|false;
```

File: src/PdoInterface.php [hunk 6]
```diff
@@ -70,18 +70,18 @@ public function errorInfo();
...
      * Gets a PDO attribute value.
      *
-     * @param mixed $attribute The PDO::ATTR_* constant.
+     * @param int $attribute The PDO::ATTR_* constant.
      *
      * @return mixed The value for the attribute.
      *
      */
-    public function getAttribute($attribute);
+    public function getAttribute(int $attribute): mixed;
```
Comment (kenjis) [inline src/PdoInterface.php]: https://www.php.net/manual/en/pdo.getattribute.php
```suggestion
    public function getAttribute(int $attribute): bool|int|string|array|null;
```

File: src/PdoInterface.php [hunk 7]
```diff
@@ -147,12 +148,12 @@ public function query($statement, ...$fetch);
...
      * @param int $parameter_type A data type hint for the database driver.
      *
-     * @return string The quoted value.
+     * @return string|false The quoted value.
      *
      * @see http://php.net/manual/en/pdo.quote.php
      *
      */
-    public function quote($value, $parameter_type = PDO::PARAM_STR);
+    public function quote(mixed $value, int $parameter_type = PDO::PARAM_STR): string|false;
```
Comment (kenjis) [inline src/PdoInterface.php]:
```suggestion
    public function quote(string|int|array|float|null $value, int $type = PDO::PARAM_STR): string|false;
```

File: src/PdoInterface.php [hunk 8]
```diff
@@ -147,12 +148,12 @@ public function query($statement, ...$fetch);
      *
      * @param int $parameter_type A data type hint for the database driver.
```
Comment (kenjis) [inline src/PdoInterface.php]:
```suggestion
     * @param int $type A data type hint for the database driver.
```

File: src/PdoInterface.php [hunk 9]
```diff
@@ -117,27 +117,28 @@ public function lastInsertId($name = null);
...
+     * @param mixed ...$fetch_mode_args Optional fetch-related parameters.
      *
-     * @return \PDOStatement
+     * @return \PDOStatement|false
      *
      * @see http://php.net/manual/en/pdo.query.php
      *
      */
-    public function query($statement, ...$fetch);
+    public function query($statement, $fetchMode = null, ...$fetch_mode_args): PDOStatement|false;
```
Comment (kenjis) [inline src/PdoInterface.php]:
```suggestion
    public function query(string $query, ?int $fetchMode = null, ...$fetch_mode_args): PDOStatement|false;
```

File: src/PdoInterface.php [hunk 10]
```diff
@@ -117,27 +117,28 @@ public function lastInsertId($name = null);
...
-     *
      */
-    public function prepare($statement, $options = null);
+    public function prepare(string $statement, array $options = []): PDOStatement|false;
 
     /**
      *
      * Queries the database and returns a PDOStatement.
      *
      * @param string $statement The SQL statement to prepare and execute.
```
Comment (kenjis) [inline src/PdoInterface.php]:
```suggestion
     * @param string $query The SQL statement to prepare and execute.
```

File: src/PdoInterface.php [hunk 11]
```diff
@@ -117,27 +117,28 @@ public function lastInsertId($name = null);
...
      * PDOStatement.
      *
-     * @return \PDOStatement
+     * @return \PDOStatement|false
      *
      * @see http://php.net/manual/en/pdo.prepare.php
-     *
      */
-    public function prepare($statement, $options = null);
+    public function prepare(string $statement, array $options = []): PDOStatement|false;
```
Comment (kenjis) [inline src/PdoInterface.php]:
```suggestion
    public function prepare(string $query, array $options = []): PDOStatement|false;
```

File: src/AbstractExtendedPdo.php [hunk 12]
```diff
@@ -574,7 +572,7 @@ public function perform($statement, array $values = [])
      * @see http://php.net/manual/en/pdo.prepare.php
      *
      */
-    public function prepare($statement, $options = [])
+    public function prepare(string $statement, array $options = []): PDOStatement
```
Comment (kenjis) [inline src/AbstractExtendedPdo.php]:
```suggestion
    public function prepare(string $query, array $options = []): PDOStatement
```

File: src/AbstractExtendedPdo.php [hunk 13]
```diff
@@ -574,7 +572,7 @@ public function perform($statement, array $values = [])
...
     {
         $this->connect();
         $sth = $this->pdo->prepare($statement, $options);
```
Comment (kenjis) [inline src/AbstractExtendedPdo.php]:
```suggestion
        $sth = $this->pdo->prepare($query, $options);
```

File: src/AbstractExtendedPdo.php [hunk 14]
```diff
@@ -635,18 +634,20 @@ public function prepareWithValues($statement, array $values = [])
...
+     *
+     * @param mixed ...$fetch_mode_args Optional fetch-related parameters.
      *
      * @return PDOStatement
      *
      * @see http://php.net/manual/en/pdo.query.php
      *
      */
-    public function query($statement, ...$fetch)
+    public function query($statement, $fetchMode = null, mixed ...$fetch_mode_args): PDOStatement
```
Comment (kenjis) [inline src/AbstractExtendedPdo.php]:
```suggestion
    public function query(string $query, ?int $fetchMode = null, mixed ...$fetch_mode_args): PDOStatement
```

File: src/AbstractExtendedPdo.php [hunk 15]
```diff
@@ -635,18 +634,20 @@ public function prepareWithValues($statement, array $values = [])
...
     {
         $this->connect();
         $this->profiler->start(__FUNCTION__);
-        $sth = $this->pdo->query($statement, ...$fetch);
+        $sth = $this->pdo->query($statement, $fetchMode, ...$fetch_mode_args);
```
Comment (kenjis) [inline src/AbstractExtendedPdo.php]:
```suggestion
        $sth = $this->pdo->query($query, $fetchMode, ...$fetch_mode_args);
```

File: src/AbstractExtendedPdo.php [hunk 16]
```diff
@@ -667,10 +668,12 @@ public function query($statement, ...$fetch)
      * @see http://php.net/manual/en/pdo.quote.php
      *
      */
-    public function quote($value, $type = self::PARAM_STR)
+    public function quote(mixed $value, int $type = self::PARAM_STR): string
```
Comment (kenjis) [inline src/AbstractExtendedPdo.php]:
```suggestion
    public function quote(string|int|array|float|null $value, int $type = self::PARAM_STR): string
```

File: src/AbstractExtendedPdo.php [hunk 17]
```diff
@@ -987,7 +990,7 @@ protected function setQuoteName($driver)
      * @param int $attribute
      * @return mixed
      */
-    public function getAttribute($attribute)
+    public function getAttribute(int $attribute): mixed
```
Comment (kenjis) [inline src/AbstractExtendedPdo.php]:
```suggestion
    public function getAttribute(int $attribute): bool|int|string|array|null
```

File: src/AbstractExtendedPdo.php [hunk 18]
```diff
@@ -1001,7 +1004,7 @@ public function getAttribute($attribute)
      * @param mixed $value
      * @return bool
      */
-    public function setAttribute($attribute, $value)
+    public function setAttribute(int $attribute, mixed$value): bool
```
Comment (kenjis) [inline src/AbstractExtendedPdo.php]:
```suggestion
    public function setAttribute(int $attribute, mixed $value): bool
```

**Comment by kenjis** [comment]
Could you update the config file of PHPUnit?

```
PHPUnit 9.5.10 by Sebastian Bergmann and contributors.

  Warning - The configuration file did not pass validation!
  The following problems have been detected:

  Line 3:
  - Element 'testsuite': The attribute 'name' is required but missing.

  Line 7:
  - Element 'filter': This element is not expected.

  Test results may not be as expected.
```

File: src/PdoInterface.php (hunk 4 above)
Comment (srjlewis) [reply src/PdoInterface.php]: I think I commented it out as static analysis showed it was not used and no tests was falling. I will be adding it back with a corresponding test.

**Comment by kenjis** [comment]
We tried to match the PDO method signatures in the PHP manual as much as possible.

**Comment by kenjis** [comment]
@srjlewis Thank you!

**Comment by kenjis** [comment]
Close #192

**Pull request merged**